src/libpmemfile-posix/read.c

```diff
@@ -128,87 +127,80 @@ pmemfile_preadv_internal(PMEMfilepool *pfp,
-	 * another thread may have destroyed the block tree while we weren't
-	 * holding the lock).
-	 */
-	os_rwlock_rdlock(&vinode->rwlock);
-	while (!vinode->blocks) {
-		os_rwlock_unlock(&vinode->rwlock);
-		os_rwlock_wrlock(&vinode->rwlock);
-
-		int err = 0;
-		if (!vinode->blocks)
-			err = vinode_rebuild_block_tree(pfp, vinode);
-		os_rwlock_unlock(&vinode->rwlock);
-
-		if (err) {
-			errno = err;
-			return -1;
-		}
-
-		os_rwlock_rdlock(&vinode->rwlock);
-	}
+	return 0;
+}
 
-	size_t ret = 0;
+static pmemfile_ssize_t
+pmemfile_preadv_internal(PMEMfilepool *pfp,
+		struct pmemfile_vinode *vinode,
+		struct pmemfile_block_desc **last_block,
+		size_t offset,
+		const pmemfile_iovec_t *iov,
+		int iovcnt)
+{
+	pmemfile_ssize_t ret = 0;
 
 	for (int i = 0; i < iovcnt; ++i) {
 		size_t len = iov[i].iov_len;
-		if ((pmemfile_ssize_t)(ret + len) < 0)
-			len = SSIZE_MAX - ret;
-		ASSERT((pmemfile_ssize_t)(ret + len) >= 0);
+		if ((pmemfile_ssize_t)((size_t)ret + len) < 0)
+			len = (size_t)(SSIZE_MAX - ret);
+		ASSERT((pmemfile_ssize_t)((size_t)ret + len) >= 0);
 
 		size_t bytes_read = vinode_read(pfp, vinode, offset, last_block,
 				iov[i].iov_base, len);
 
-		ret += bytes_read;
+		ret += (pmemfile_ssize_t)bytes_read;
 		offset += bytes_read;
 		if (bytes_read != len)
 			break;
 	}
 
-	bool update_atime = !(file_flags & PFILE_NOATIME);
+	return ret;
+}
+
+/*
+ * handle_atime -
+ * Updates the atime field following a read operation, if necessary.
+ * The vinode must not be locked when calling this function.
+ */
+static void
+handle_atime(PMEMfilepool *pfp,
+		struct pmemfile_vinode *vinode,
+		uint64_t file_flags)
+{
+	if (file_flags & PFILE_NOATIME)
+		return;
+
+	struct pmemfile_inode *inode = vinode->inode;
 	struct pmemfile_time tm;
 
-	if (update_atime) {
-		struct pmemfile_time tm1d;
-		if (get_current_time(&tm)) {
-			LOG(LINF, "can not get current time");
-			update_atime = false;
-		} else {
-			tm1d.nsec = tm.nsec;
-			tm1d.sec = tm.sec - 86400;
-
-			/* relatime */
-			update_atime =	time_cmp(&inode->atime, &tm1d) < 0 ||
-				time_cmp(&inode->atime, &inode->ctime) < 0 ||
-				time_cmp(&inode->atime, &inode->mtime) < 0;
-		}
+	if (get_current_time(&tm)) {
+		LOG(LINF, "can not get current time");
+		return;
 	}
 
-	os_rwlock_unlock(&vinode->rwlock);
+	struct pmemfile_time tm1d = tm;
+	tm1d.sec -= 86400;
+
+	/* relatime */
+	if ((time_cmp(&inode->atime, &tm1d) >= 0) &&
+	    (time_cmp(&inode->atime, &inode->ctime) >= 0) &&
+	    (time_cmp(&inode->atime, &inode->mtime) >= 0))
+		return;
 
 	ASSERT_NOT_IN_TX();
-	if (update_atime) {
-		os_rwlock_wrlock(&vinode->rwlock);
 
-		TX_BEGIN_CB(pfp->pop, cb_queue, pfp) {
-			TX_SET_DIRECT(inode, atime, tm);
-		} TX_ONABORT {
-			LOG(LINF, "can not update inode atime");
-		} TX_END
+	os_rwlock_wrlock(&vinode->rwlock);
 
-		os_rwlock_unlock(&vinode->rwlock);
-	}
+	TX_BEGIN_CB(pfp->pop, cb_queue, pfp) {
+		vinode->metadata_modification_counter++;
+		memory_barrier();
```

Comment:
Maybe.
Note: I like showing this explicitly in the source here. This is rather meaningful if/when we use the modification_counters to detect changes without locking the vinode:
```
modifying_routine(vinode)
{
    vinode->metadata_modification_counter++;
    memory_barrier();
    ...perform modifications...
}

reading_routine(vinode)
{
    ...perform reads...
    memory_barrier();
    if (is_modification_indicated(vinode)) {
       lock(vinode)
        ...perform reads...
       unlock(vinode)
    }    
}
```